src/plugins/redirections/crossing/checkCrossingRedirections.ml

```diff
@@ -0,0 +1,232 @@
+    descriptors;
+  print_newline ()
+
+let have_descriptors_crossed descriptors =
+  let result = ref None in
+  for i = 0 to Array.length descriptors -1 do
+    match descriptors.(i) with
+    | Initial j when j <> i ->
+       (
+         match descriptors.(j) with
+         | Initial k when k = j -> ()
+         | _ ->
+            (* i points to what j used to point to; but j points to
+               something else => this is what we wanted to detect *)
+            result := Some (i, j)
+       )
+    | _ -> ()
+  done;
+  !result
+
+let check_io_redirect_list io_redirect_list =
+  let descriptors = initial_descriptors () in
+  apply_io_redirect_list descriptors io_redirect_list;
+  match have_descriptors_crossed descriptors with
+  | None -> []
+  | Some (i, j) ->
+     [Alarm.make
+        ~position:(List.hd io_redirect_list).position
+        (Printf.sprintf "The file descriptor %d points to what %d used to point to; but %d points to something else. You probably want to change the order of your redirections." i j j)]
+
+(* ========================= [ The checker itself ] ========================= *)
+
+module Checker : Analyzer.S = struct
+  let name = "redirections/crossing"
+  let documentation = "Looks for wrong redirections like 2>&1 >/dev/null"
+
+  let analyzer (csts: complete_command_list) =
+    let visitor = object (self)
+      inherit [_] Libmorbig.CST.reduce as super
+
+      method zero = []
+      method plus = (@)
+
+      method! visit_command () command =
+        (
+          match command with
+          | Command_CompoundCommand_RedirectList (_, redirect_list') ->
+             (
+               redirect_list_to_io_redirect_list redirect_list'
+               |> check_io_redirect_list
+             )
+          | _ -> self#zero
+        )
+        |> self#plus (super#visit_command () command)
+
+      method! visit_function_body () function_body =
+        (
+          match function_body with
+          | FunctionBody_CompoundCommand_RedirectList (_, redirect_list') ->
+             (
+               redirect_list_to_io_redirect_list redirect_list'
+               |> check_io_redirect_list
+             )
+          | _ -> self#zero
+        )
+        |> self#plus (super#visit_function_body () function_body)
+
+      method! visit_simple_command () simple_command =
+        (
+          match simple_command with
+          | SimpleCommand_CmdPrefix_CmdWord_CmdSuffix (cmd_prefix', _, cmd_suffix') ->
+             (
+               let in_prefix = cmd_prefix_to_io_redirect_list cmd_prefix' in
+               let in_suffix = cmd_suffix_to_io_redirect_list cmd_suffix' in
+               match in_prefix, in_suffix with
+               | [], [] -> []
+               | _, [] -> check_io_redirect_list in_prefix
+               | [], _ -> check_io_redirect_list in_suffix
+               | _, _ -> check_io_redirect_list (in_prefix @ in_suffix)
+             )
+          | SimpleCommand_CmdPrefix_CmdWord (cmd_prefix', _) ->
```

Comment:
There are some redundancies in the next methods implementations, right?

Reply:
In fact, the whole function that takes a simple command and returns a list of redirections could be externalized in Morbig's CST helpers. It makes sense as it will be used in shstats, lintshell and morsmall.

Reply:
Yes. (I will not be surprised to see a PR coming on the Morbig's side :)

Reply:
(incoming)